src/work/PokerGame/PokerGame.sol

```diff
@@ -0,0 +1,516 @@
+        }
+
+        currentPlayer.lastAction = (amount > game.highestBet) ? PlayerAction.RAISE : PlayerAction.CALL; // Set the player's last action
+
+        emit PlayerAction(gameId, msg.sender, currentPlayer.lastAction, amount); // Emit the PlayerAction event
+
+        nextTurn(gameId); // Move to the next turn
+    }
+
+    // Function for a player to fold
+    function fold(uint256 gameId) external {
+        Game storage game = games[gameId]; // Get the game from storage
+        require(game.state == GameState.ACTIVE, "Game not active"); // Ensure the game is in the ACTIVE state
+        require(msg.sender == game.players[game.currentPlayerIndex].addr, "Not your turn"); // Ensure it is the player's turn
+
+        Player storage currentPlayer = game.players[game.currentPlayerIndex]; // Get the current player
+        currentPlayer.isActive = false; // Set the player as inactive
+        currentPlayer.lastAction = PlayerAction.FOLD; // Set the player's last action to FOLD
+
+        emit PlayerAction(gameId, msg.sender, PlayerAction.FOLD, 0); // Emit the PlayerAction event
+
+        nextTurn(gameId); // Move to the next turn
+    }
+
+    // Internal function to move to the next turn
+    function nextTurn(uint256 gameId) internal {
+        Game storage game = games[gameId]; // Get the game from storage
+            Player storage currentPlayer = game.players[game.currentPlayerIndex];
+
+        _revokeRole(CURRENT_TURN_ROLE, currentPlayer.addr);
+        currentPlayer.hasCurrentTurnRole = false;
+
+        uint8 nextPlayerIndex = (game.currentPlayerIndex + 1) % game.playerCount; // Calculate the next player index
+
+        while (!game.players[nextPlayerIndex].isActive) { // Find the next active player
+            nextPlayerIndex = (nextPlayerIndex + 1) % game.playerCount;
+        }
+
+        game.currentPlayerIndex = nextPlayerIndex; // Set the current player index to the next active player
+        Player storage nextPlayer = game.players[nextPlayerIndex];
+
+        _grantRole(CURRENT_TURN_ROLE, nextPlayer.addr);
+        nextPlayer.hasCurrentTurnRole = true;
+
+        if (isRoundComplete(gameId)) {
+            // Handle end of betting round logic here
+            // This could involve dealing community cards, starting a new betting round, or ending the game
+        }
+
+        emit TurnChanged(gameId, game.players[game.currentPlayerIndex].addr); // Emit the TurnChanged event
+    }
+
+    // Internal function to check if the betting round is complete
+    function isRoundComplete(uint256 gameId) internal view returns (bool) {
+        Game storage game = games[gameId]; // Get the game from storage
+        uint256 activePlayers = 0; // Count of active players
+        uint256 playersActed = 0; // Count of players who have acted
+        uint256 highestBet = 0; // Highest bet in the current round
+
+        for (uint8 i = 0; i < game.playerCount; i++) { // Loop through all players
```

Comment:
avoid running these loops if you can. ie use a counter to keep track of number of players that have acted during the round